src/ray/core_worker/reference_count.h

```diff
@@ -587,89 +664,71 @@ class ReferenceCounter : public ReferenceCounterInterface,
-    /// ObjectRefs nested in this object that are or were in use. These objects
-    /// are not owned by us, and we need to report that we are borrowing them
-    /// to their owner. Nesting is transitive, so this flag is set as long as
-    /// any child object is in scope.
-    bool has_nested_refs_to_report = false;
-    /// A list of processes that are we gave a reference to that are still
-    /// borrowing the ID. This field is updated in 2 cases:
-    ///  1. If we are a borrower of the ID, then we add a process to this list
-    ///     if we passed that process a copy of the ID via task submission and
-    ///     the process is still using the ID by the time it finishes its task.
-    ///     Borrowers are removed from the list when we recursively merge our
-    ///     list into the owner.
-    ///  2. If we are the owner of the ID, then either the above case, or when
-    ///     we hear from a borrower that it has passed the ID to other
-    ///     borrowers. A borrower is removed from the list when it responds
-    ///     that it is no longer using the reference.
-    absl::flat_hash_set<rpc::WorkerAddress> borrowers;
-    /// When a process that is borrowing an object ID stores the ID inside the
-    /// return value of a task that it executes, the caller of the task is also
-    /// considered a borrower for as long as its reference to the task's return
-    /// ID stays in scope. Thus, the borrower must notify the owner that the
-    /// task's caller is also a borrower. The key is the task's return ID, and
-    /// the value is the task ID and address of the task's caller.
-    absl::flat_hash_map<ObjectID, rpc::WorkerAddress> stored_in_objects;
+    /// Whether the lineage of this object was evicted due to memory pressure.
+    bool lineage_evicted = false;
     /// The number of tasks that depend on this object that may be retried in
     /// the future (pending execution or finished but retryable). If the object
     /// is inlined (not stored in plasma), then its lineage ref count is 0
     /// because any dependent task will already have the value of the object.
     size_t lineage_ref_count = 0;
-    /// Whether the lineage of this object was evicted due to memory pressure.
-    bool lineage_evicted = false;
-    /// Whether this object has been spilled to external storage.
-    bool spilled = false;
+
+    /// The local ref count for the ObjectID in the language frontend.
+    size_t local_ref_count = 0;
+    /// The ref count for submitted tasks that depend on the ObjectID.
+    size_t submitted_task_ref_count = 0;
+
+    /// Metadata for references that contain this reference.
+    std::unique_ptr<ContainingReferences> containing_references;
+
+    /// Metadata related to borrowing.
+    std::unique_ptr<BorrowInfo> borrow_info;
+
+    /// Callback that will be called when this ObjectID no longer has
+    /// references.
+    std::function<void(const ObjectID &)> on_delete;
```

Comment:
We can do it in a different PR, but out of curiosity, did you try testing what would happen if we use the same callback for all objects instead of saving them individually?